src/benchmarks/micro/corefx/System.Collections/Sort.cs

```diff
@@ -11,34 +11,47 @@
 namespace System.Collections
 {
     [BenchmarkCategory(Categories.CoreCLR, Categories.Collections, Categories.GenericCollections)]
-    [GenericTypeArguments(typeof(int))] // value type
-    [GenericTypeArguments(typeof(string))] // reference type
+    [GenericTypeArguments(typeof(int))] // value type, Array sort in native code
+    [GenericTypeArguments(typeof(IntStruct))] // custom value type, sort in managed code
+    [GenericTypeArguments(typeof(IntClass))] // custom reference type, sort in managed code, compare fast
+    [GenericTypeArguments(typeof(BigStruct))] // custom value type, sort in managed code
+    [GenericTypeArguments(typeof(string))] // reference type, compare slow
     [InvocationCount(InvocationsPerIteration)]
-    public class Sort<T>
+    public class Sort<T> where T : IComparable<T>
     {
+        static readonly ComparableComparerClass _comparableComparerClass = new ComparableComparerClass();
         private const int InvocationsPerIteration = 1000;
 
         [Params(Utils.DefaultCollectionSize)]
         public int Size;
 
         private int _iterationIndex = 0;
         private T[] _values;
-        
         private T[][] _arrays;
         private List<T>[] _lists;
 
         [GlobalSetup]
-        public void Setup() => _values = ValuesGenerator.ArrayOfUniqueValues<T>(Size);
-        
+        public void Setup() => _values = GenerateValues();
+
         [IterationCleanup]
         public void CleanupIteration() => _iterationIndex = 0; // after every iteration end we set the index to 0
```

Comment:
@billwert @jkotas  I think there is a bug in the List benchmark, the `CleanupIteration` should reset `_lists = null` since otherwise, the Lists will be used, that is how `Utils.FillCollections` work it seems, unlike `UtilsFillArray`, the way `Utils.FillCollections` is a bit prone to errors for tests that are "destructive".

Reply:
@nietras Thanks for finding and fixing the bug!

The goal of the method executed as `IterationSetup` is to take less time than the benchmarked method itself. If we allocate a new list and add all the elements to it it's going to take more time than Sort itself. When we run such a benchmark it will be harder to get stable results.

Could you please fix this bug by not setting the `list` to `null` in the IterationCleanup but instead implementing a new method that will just put the elements using indexer? Sth like:

```cs
internal static void SetItems<TCollection, TValues>(ref TCollection[] collections, int collectionsCount, TValues[] keys)
    where TCollection : IList<TValues>, new()
{
    if (collections == null)
        collections = Enumerable.Range(0, collectionsCount).Select(_ => new TCollection()).ToArray();

    foreach (var collection in collections)
        for (int i = 0; i < keys.Length; i++)
            collection[i] = keys[i];

    if (collections.Any(collection => collection.Count != keys.Length)) // we dont use Debug.Assert here because this code will be executed mostly in Release
        throw new InvalidOperationException();
}
```

Thanks!

Reply:
@adamsitnik no problem, but won't what you suggest fail if we get more size test e.g.:
```
        [Params(512, 256)]
        public int Size;
```
If `IterationSetup` should be fast, why not allocated these before that? 

The other thing is that the `Utils` helper methods seem prone to errors e.g.:
```csharp
internal static void FillCollections<TCollection, TValues>(ref TCollection[] collections, int collectionsCount, TValues[] keys)
    where TCollection : ICollection<TValues>, new()
{
    if (collections == null)
        collections = Enumerable.Range(0, collectionsCount).Select(_ => new TCollection()).ToArray();

    foreach (var collection in collections.Where(collection => collection.Count < keys.Length))
    foreach (var value in keys)
        collection.Add(value);
    
    if(collections.Any(collection => collection.Count != keys.Length)) // we dont use Debug.Assert here because this code will be executed mostly in Release
        throw new InvalidOperationException();
}
```
If you change `collectionsCount` during benchmark e.g. from 100 to 10, this will not fail and you will have more collections than "asked for". Depending on the benchmark this could be a problem.

Reply:
@adamsitnik  the main source of variance here is in fact that `InvocationsPerIteration` should be different depending on `T` although this of course gets a separate run per `T`. In some ways, the code implicitly depends a lot on how `BenchmarkDotNet` executes the benchmark, which may be ok of course.

Reply:
@adamsitnik I have added `ClearAndFillCollections` instead, this ensure the length always matches the `TValues[]` given.

Reply:
@adamsitnik also added check for `collectionsCount` let me know if that is ok :)